### PR TITLE
generic conn pool: directly use thread local cluster

### PR DESCRIPTION
**Pull request by mattklein123**
Further cleanup from previous PRs. In the happy path this further
eliminates map lookups and simplifies error handling.

Risk Level: Low
Testing: Existing tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: test/extensions/upstreams/tcp/generic/config_test.cc [hunk 1]
```diff
@@ -1,39 +0,0 @@
...
-  const std::string cluster_name_{"cluster_name"};
-  GenericConnPoolFactory factory_;
-  NiceMock<Envoy::Tcp::ConnectionPool::MockUpstreamCallbacks> callbacks_;
-};
-
-TEST_F(TcpConnPoolTest, TestNoMatchingClusterName) {
-  envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig config;
-  config.set_hostname("host");
-  EXPECT_CALL(cluster_manager_, getThreadLocalCluster(_)).WillOnce(Return(nullptr));
-  EXPECT_EQ(nullptr, factory_.createGenericConnPool(cluster_name_, cluster_manager_, config,
```
Comment (alyssawilk) [inline test/extensions/upstreams/tcp/generic/config_test.cc:31]: Could instead have the thread local cluster return null for the tcpConnPool call?  The goal was to test the factory returning null which I think can still happen.

**Review by alyssawilk — COMMENTED** [review]
Nice cleanup!  LG, modulo the one test nit.

File: test/extensions/upstreams/tcp/generic/config_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/extensions/upstreams/tcp/generic/config_test.cc:31]: OK got it. I will bring this back.

File: test/extensions/upstreams/tcp/generic/config_test.cc [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+class TcpConnPoolTest : public ::testing::Test {
+public:
+  NiceMock<Upstream::MockThreadLocalCluster> thread_local_cluster_;
+  GenericConnPoolFactory factory_;
+  NiceMock<Envoy::Tcp::ConnectionPool::MockUpstreamCallbacks> callbacks_;
+};
+
+TEST_F(TcpConnPoolTest, TestNoHost) {
+  envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_TunnelingConfig config;
```
Comment (alyssawilk) [inline test/extensions/upstreams/tcp/generic/config_test.cc:27]: Arguably there is a host, just no connection to it, but that's super nitty so feel free to merge in as is :-)

**Review by alyssawilk — DISMISSED** [review]

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**